### PR TITLE
[Draft] Add a `Pipeline.evaluate` method

### PR DESCRIPTION
**Pull request by dcfidalgo**
This PR adds an `evaluate` method to our `Pipeline` class addressing issue #406 . I left some todos, since i want to have a quick discussion first, before considering this for a merge.

Right now the `Pipeline` class has the `predict`/`explain`/`evaluate` methods that only really make sense for the `_PreTrainedPipeline` class. To make a meaningful prediction you have following flow:

- `pl = Pipeline.from_config()` -> `pl.train()` -> `pl = Pipeline.from_pretrained()` -> `pl.predict` /`pl.evaluate`

The reason is that we do not modify the weights "in-place", but create a copy of the pipeline when we train.
The **advantages** of an "in-place" modification would be:
- allows for a straight forward flow: `pl.from_config` -> `pl.train` -> `pl.predict`/`pl.evaluate` (right now, this flow will "work", but with an unexpected result);
- less memory footprint, no need to keep two models in memory when training;

**Disadvantages** would be (maybe i am missing some!):
- consecutive trainings from scratch would need an intermediate reset step: `pl.from_config()` -> `pl.train` -> `pl.reset` -> `pl.train`

In summary i vote for either: 
- implementing some sort of `reset` method that allows to reset a model to its initial state after a training, and modify the weights in-place,
- or moving the inference methods to the `_PreTrainedPipeline` class, although this will break the possibility of calling the `predict` method to test a pipeline configuration.

@frascuchon what do you think?

### CODE REVIEW

**Comment by dcfidalgo** [comment]
Moving the discussion from #411 to here.

In #411 @frascuchon suggested a very nice idea. We could have a context manager that restores the model state on exit.

So if you want to do consecutive trainings from scratch you could do something like this:
```python
with pl.restore_model():
    pl.train()
```

**Comment by frascuchon** [comment]
@dcfidalgo It's this PR still in draft ?

**Comment by dcfidalgo** [comment]
> @dcfidalgo It's this PR still in draft ?

I think it should work the way it is implemented in this PR. But maybe i would like to discuss the issues raised above first, before merging this.

**Comment by frascuchon** [comment]
> In summary i vote for either:
>
> implementing some sort of reset method that allows to reset a model to its initial state after a training, and modify the weights in-place,

In that case, `reset` is not strictly necessary. We can just decide mutable pipelines, indicating in documentation how to apply multiple trainings over same configuration:

````python
for train_id in range(1,10):
  pl = Pipeline.from_config(....)
  ...
  pl.train(...)
```` 

> or moving the inference methods to the _PreTrainedPipeline class, although this will break the possibility of calling the predict method to test a pipeline configuration.

We need to keep the "mock prediction" for now, at least until design a better path for data annotation sessions.


You can go ahead with this approach (mutable pipeline), but keeping predictions fro non trained pipelines.

**Comment by dvsrepo** [comment]
> > In summary i vote for either:
> > implementing some sort of reset method that allows to reset a model to its initial state after a training, and modify the weights in-place,
> 
> In that case, `reset` is not strictly necessary. We can just decide mutable pipelines, indicating in documentation how to apply multiple trainings over same configuration:
> 
> ```python
> for train_id in range(1,10):
>   pl = Pipeline.from_config(....)
>   ...
>   pl.train(...)
> ```
> 
> > or moving the inference methods to the _PreTrainedPipeline class, although this will break the possibility of calling the predict method to test a pipeline configuration.
> 
> We need to keep the "mock prediction" for now, at least until design a better path for data annotation sessions.
> 
> You can go ahead with this approach (mutable pipeline), but keeping predictions fro non trained pipelines.

👍

**Comment by dcfidalgo** [comment]
Closing this one, see #462
